### PR TITLE
[RDF] Add progress bar + event statistics

### PR DESCRIPTION
**Pull request by hageboeck**
# This Pull request:
ProgressHelper is a class that offers callback functions for RDataFrame,
and can compute and print event statistics and a progress bar.

With a change like this:
```diff
--- a/tutorials/dataframe/df102_NanoAODDimuonAnalysis.C
+++ b/tutorials/dataframe/df102_NanoAODDimuonAnalysis.C
@@ -28,6 +28,7 @@
 #include "TLatex.h"
 #include "Math/Vector4D.h"
 #include "TStyle.h"
+#include "ROOT/RDFHelpers.hxx"
 
 using namespace ROOT::VecOps;
 
@@ -52,6 +53,9 @@ void df102_NanoAODDimuonAnalysis()
    // Request cut-flow report
    auto report = df_mass.Report();
 
+   ROOT::RDF::ProgressHelper progress(10000, ROOT::RDF::CountEvents("Events", "root://eospublic.cern.ch//eos/opendata/cms/derived-data/AOD2NanoAODOutreachTool/Run2012BC_DoubleMuParked_Muons.root"));
+   h.OnPartialResultSlot(10000, [&progress](unsigned int slot, TH1D& histo){ progress(slot, histo); });
+
    // Produce plot
    gStyle->SetOptStat(0); gStyle->SetTextFont(42);
    auto c = new TCanvas("c", "", 800, 700);
```

one gets:
`bin/root -q ~/code/root-src/tutorials/dataframe/df102_NanoAODDimuonAnalysis.C+O`
image

## TODO
- [ ] Write helpers that integrate this into the head node of RDF, preferably with a single line of code.
- [ ] *Optional*: Finalise callbacks for RDF, since a carriage return is always missing.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

**Comment by jalopezg-r00t** [comment]
Niiiiceeee!  :-D

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 1]
```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
...
+  template<typename T>
+  void operator()(T& /*value*/) {
+    using namespace std::chrono;
+    // ***************************************************
+    // Warning: Here, everything needs to be thread safe:
+    // ***************************************************
+    fProcessedEvents += fIncrement;
+
+    // We only print every 1s, and only one thread does it. Try to take the lock:
+    if (duration_cast<seconds>(system_clock::now() - fLastPrintTime).count() == 0 || !fPrintMutex.try_lock())
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:238]: The access to `fLastPrintTime` and `fIncrement` seems thread unsafe?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 2]
```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
...
+      return;
+
+    // ***************************************************
+    // Protected by lock from here:
+    // ***************************************************
+    std::lock_guard<std::mutex> lockGuard(fPrintMutex, std::adopt_lock_t{});
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:244]:
```suggestion
    std::lock_guard<std::mutex> lockGuard(fPrintMutex, std::adopt_lock);
```

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 3]
```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
...
+
+    std::size_t eventCount;
+    seconds elapsedSeconds;
+    std::tie(eventCount, elapsedSeconds) = RecordEvtCountAndTime();
+
+    std::cout << "\r";
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:250]: What happens on Windows?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 4]
```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
...
+
+// clang-format off
+/// Count events in a tree in the file described by `fileUrl`.
+/// \note This will open and close the file, so don't use this in loops.
+///
+/// \param treename Name of the tree.
+/// \param fileUrl Path or URL to a ROOT file.
+/// \return Event counts. Returns 0 in case of errors.
+// clang-format on
+std::size_t CountEvents(const char* treename, const char* fileUrl);
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: I don't see this function being called?

File: tree/dataframe/src/RDFHelpers.cxx [hunk 5]
```diff
@@ -57,3 +60,124 @@ void ROOT::RDF::RunGraphs(std::vector<RResultHandle> handles)
...
+      stream << (i > 1 ? '.' : 'k');
+    }
+    return stream << std::setfill(' ');
+  }
+
+  /// Format std::chrono::seconds as `1:30m`.
+  std::ostream& operator<<(std::ostream& stream, std::chrono::seconds elapsedSeconds) {
+    stream << std::chrono::duration_cast<std::chrono::minutes>(elapsedSeconds).count()
+        << ":" << std::setw(2) << std::right << std::setfill('0') << elapsedSeconds.count() % 60;
+    return stream << 'm' << std::setfill(' ');
```
Comment (Axel-Naumann) [inline tree/dataframe/src/RDFHelpers.cxx:121]: `setfill(' ')` might not be resetting it to the *prior* state, and `setw` might need to be reset, too? See https://stackoverflow.com/questions/2273330/restore-the-state-of-stdcout-after-manipulating-it

File: tree/dataframe/src/RDFHelpers.cxx [hunk 6]
```diff
@@ -57,3 +60,124 @@ void ROOT::RDF::RunGraphs(std::vector<RResultHandle> handles)
...
+    return;
+
+  const double completion = double(currentEventCount) / fMaxEvents;
+  const unsigned int nBar = completion * fBarWidth;
+
+  std::string bars(std::max(nBar, 1u), '=');
+  bars.back() = (nBar == fBarWidth) ? '=' : '>';
+
+  if (fUseShellColours) stream << "\e[33m";
+  stream << '|' << std::setfill(' ') << std::setw(fBarWidth) << std::left << bars << "|   ";
```
Comment (Axel-Naumann) [inline tree/dataframe/src/RDFHelpers.cxx:170]: See above on `ios` resetting manipulation

File: tree/dataframe/src/RDFHelpers.cxx [hunk 7]
```diff
@@ -57,3 +60,124 @@ void ROOT::RDF::RunGraphs(std::vector<RResultHandle> handles)
    for (auto &h : uniqueLoops)
       run(h);
 }
+
+
+/// Compute a running mean of events/s.
+double ROOT::RDF::ProgressHelper::EvtPerSec() const {
```
Comment (Axel-Naumann) [inline tree/dataframe/src/RDFHelpers.cxx:66]: Do you think it'd be possible to test some of the functionality - even if it won't be the progress bar itself?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 8]
```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
 // clang-format on
 void RunGraphs(std::vector<RResultHandle> handles);
 
+
+// clang-format off
+/// RDF progress helper.
+/// This class provides callback functions to RDataFrame, which record event statistics
+/// and print them to the terminal every second.
+/// ProgressHelper::operator()(unsigned int, T&) is thread safe, and can be used as callback in MT mode.
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:192]: Can you show the actual code to be used, given an `RDataFrame rdf`?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 9]
```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
...
+  /// Create a progress helper.
+  /// \param increment RDF callbacks are called every `n` events. Pass this `n` here.
+  /// \param maxEvent If known, pass the number of events in the dataset. This will allow for computing completion
+  /// statistics.
+  /// \param progressBarWidth Number of characters the progress bar will occupy.
+  /// \param useShellColours Use shell colour codes to colour the output.
+  ProgressHelper(std::size_t increment,
+      std::size_t maxEvent,
+      unsigned int progressBarWidth = 40,
+      bool useShellColours = true) :
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:206]: Can you initialize this to `false` if Windows or `!isatty(fileno(stdout))`?

**Review by Axel-Naumann — COMMENTED** [review]
Super nice! Just a couple of questions :-)

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 9 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:206]: Tell me how to detect that, and yes!
It's configurable because it should be, but I don't know how to auto-detect.

Also, we cannot use carriage return when writing to a file, so you see funny stuff in that file. I'd argue that it shouldn't be used when writing to files or we need a non-interactive mode.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 1 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:238]: `const std::size_t fIncrement;`
I will say no more.

`fLastPrintTime` might indeed change, but the logic is this:
- If the time is right, try take a lock.
- If it's not right: exit.
- If it's right, but you cannot take a lock: exit.

Logic flawed? The progress bar behaved very nicely in my tests. Thread sanitizer might complain, though.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 3 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:250]: As a first step: Don't use it on windows.
Second step: I don't know. What should happen?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 8 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:192]: Code not finalised yet (see TODOs), but yes!

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 4 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: Not here, but see code example in PR description. It's in the TODOs that we need to decide how to make this a one-liner.
- [ ] We also need to point out that we are accessing each file *once more*, so maybe extra connections to handle.
- [ ] We also need a solution for chains or trees that come via pointer, e.g. an overload.

File: tree/dataframe/src/RDFHelpers.cxx (hunk 7 above)
Comment (hageboeck) [reply tree/dataframe/src/RDFHelpers.cxx:66]: Hmm, possible sure. I didn't think about tests much yet, though.

File: tree/dataframe/src/RDFHelpers.cxx (hunk 5 above)
Comment (hageboeck) [reply tree/dataframe/src/RDFHelpers.cxx:121]: 👏

Comment (hageboeck) [reply tree/dataframe/src/RDFHelpers.cxx:121]: @Axel-Naumann `fmtflags` is not saving the fill character, is it?

**Comment by eguiraud** [comment]
Thanks a lot Stephan!
On the line of your first to-do item, I think the programming model should be:

```cpp
auto progbar = ROOT::RDF::MakeProgressBar(df, everyNEvents, nEvents); // last 2 args are optional
```

rather than

```cpp
ROOT::RDF::ProgressHelper progress(everyNEvents, nEvents);
h.OnPartialResultSlot(10000, [&progress](unsigned int slot, TH1D& histo){ progress(slot, histo); });
```

I would also rename `CountEvents` to the slightly scarier `RetrieveNEvents` (which might take string_views rather than `const char *`) to hint at the fact that it can be an expensive call.

File: tree/dataframe/src/RDFHelpers.cxx (hunk 5 above)
Comment (hageboeck) [reply tree/dataframe/src/RDFHelpers.cxx:121]: I don't think so, but I fixed that.

**Comment by hageboeck** [comment]
> Thanks a lot Stephan!
> On the line of your first to-do item, I think the programming model should be:
> 
> ```c++
> auto progbar = ROOT::RDF::MakeProgressBar(df, everyNEvents, nEvents); // last 2 args are optional
> ```
> 
> rather than
> 
> ```c++
> ROOT::RDF::ProgressHelper progress(everyNEvents, nEvents);
> h.OnPartialResultSlot(10000, [&progress](unsigned int slot, TH1D& histo){ progress(slot, histo); });
> ```
> 
> I would also rename `CountEvents` to the slightly scarier `RetrieveNEvents` (which might take string_views rather than `const char *`) to hint at the fact that it can be an expensive call.

Excellent!

Tell me though: An RDF (= RNode I think?) doesn't take callbacks. I need a `RResultPtr`, correct? Do I have to attach a dummy result or did I miss something?

**Comment by hageboeck** [comment]
> ```c++
> auto progbar = ROOT::RDF::MakeProgressBar(df, everyNEvents, nEvents); // last 2 args are optional
> ```

Although the ProgressBar won't be a progress bar when we don't know the number of events ...

**Comment by hageboeck** [comment]
> Tell me though: An RDF (= RNode I think?) doesn't take callbacks. I need a `RResultPtr`, correct? Do I have to attach a dummy result or did I miss something?

OK, I had a look, and it would actually be amazing if I could get access to the LoopManager, because then I could just register the callback using `RLoopManager::RegisterCallback` (cannot link to docs because undocumented functions don't get an anchor in doxygen, any more 🤷‍♂️ ).
I don't see a way to get to the [LoopManager](https://root.cern.ch/doc/master/classROOT_1_1Detail_1_1RDF_1_1RLoopManager.html), though.

**Comment by eguiraud** [comment]
> Although the ProgressBar won't be a progress bar when we don't know the number of events ...

You're right, better name pending (`AddProgressIndicator`?)

> It would actually be amazing if I could get access to the LoopManager, because then I could just register the callback using RLoopManager::RegisterCallback

Yes that's a great idea, there are actually several advantages:

- as you mention, in the callback you wouldn't have to discard a partially evaluated result anymore
- `AddProgressIndicator` wouldn't need to return anything, it can just do what it says without users having to deal with the returned object
- probably the biggest: we wouldn't have to do any magic to make sure that the method is called _on the head node_ (because if you call it on a `Filter`, now you are updating the progress bar every N _filtered_ entries) 

Now -- of course we don't just expose the `RLoopManager` to _anyone_, but we trust you, so you can add a `ROOT::Internal::ExtractLoopManager` function as a friend of `RInterface` that just returns `interface.fLoopManager`?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 1 above)
Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:238]: > `const std::size_t fIncrement;` I will say no more.

:+1:

> Logic flawed?

LGTM, `RecordEvtCountAndTime` is only ever called from the locked region.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 3 above)
Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:250]: What should happen? A nice progress bra is displayed, or at least the terminal isn't flooded with new lines with some number of `=` :-)

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 4 above)
Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: I see, I suppose that indeed needs some discussion. We should strive for a solution that takes estimates:
- all files have the same amount of entries, open one and scale with the number of files according to RDF
- as the callback is invoked on a newly opened file, update the assumption

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 9 above)
Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:206]: Can you first find out what happens on Windows, even with `useShellColours = false`?

Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:206]: Oh, and we spell that as `useShellColors`, I guess in recognition of ROOT's first user collaborations, which were at Fermilab ;-)

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 4 above)
Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: @eguiraud ^

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: To do something like that we'd need to make the progress bar a core RDF feature (and then add logic to update the guesstimate when a new file is opened, only for datasources that support multiple input files...).

In order to keep things simple, I am fine with just printing `XXX evts/s, YYY evts processed`, without a progress bar, if the total number of events is not provided.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 3 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:250]: Yes, yes, sure. Let me rephrase the question:
What do we have to do to get something working on windows?
I don't know how their terminals work, and I don't have the time to look it up.

And why do you want a
> bra
to be displayed? Can we settle for a bar as a first step? 😄

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 4 above)
Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: Virtually everybody (not using distRDF) will want this feature. Asking users to provide the total event count will trigger the following:
- tutorials will pass the parameter
- analyses will pass the parameter, causing all input files to be opened
- repeat for the next 20 years, with enormous integrated perf losses.

IMHO a progress bar *is* a standard RDF feature, and we need to reduce the damage. I don't think the design you describe is the only option; I could imaging datasources providing abstracted access to "input identifier", "number of entries", such that the progress bar can check whether its datasource identifier has changes, and if so, update the number of entries. Or RDF could construct that progress bar with a callback, or at least with the number of dataset elements - which also sounds like a useful interface for datasources.

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: I was describing requirements more than a design.

What you describe is a different kind of effort than a helper function. Please open a feature request and I'll get to it as soon as possible :)

Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: That's missing a point: if we add this, people will use it and it will have detrimental effects on performance. Even though we won't get to doing the better option in a reasonable time we should not add a solution for which we predict detrimental effects, even if that option requires way less effort.

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: I didn't miss that point :)

Comment (Axel-Naumann) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:289]: Oh okay sorry - I didn't realize that. I've opened the [feature request](https://github.com/root-project/root/issues/8685) you asked for.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 9 above)
Comment (hageboeck) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx]: 👍 for tty. How could we test this on windows? @bellenot

Comment (bellenot) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx]: @hageboeck I'll have to look at the implementation

File: tree/dataframe/src/RDFHelpers.cxx [hunk 10]
```diff
@@ -57,3 +62,156 @@ void ROOT::RDF::RunGraphs(std::vector<RResultHandle> handles)
...
+  if (fUseShellColours) stream << "\e[0m";
+
+
+  // events/s
+  stream << std::scientific << std::setprecision(2) << evtpersec << " evt/s";
+
+
+  // Time statistics:
+  if (fMaxEvents != 0) {
+    if (fUseShellColours) stream << "\e[35m";
```
Comment (bellenot) [inline tree/dataframe/src/RDFHelpers.cxx:182]: This kind of formatting is indeed not going to work on Windows...

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 9 above)
Comment (bellenot) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx]: @hageboeck after a quick look at the code, I confirm that `useShellColours` must be `false` on Windows

File: tree/dataframe/src/RDFHelpers.cxx (hunk 10 above)
Comment (hageboeck) [reply tree/dataframe/src/RDFHelpers.cxx:182]: `#ifdef _WIN32`?

What about `\r` for a carriage return?

Comment (bellenot) [reply tree/dataframe/src/RDFHelpers.cxx:182]: everything like `"\e[35m"` will not work, so `fUseShellColours` must simply be false on Windows

**Comment by bellenot** [comment]
Note that we can use colors in the Windows terminal, the way to do it is just different

File: tree/dataframe/src/RDFHelpers.cxx (hunk 10 above)
Comment (hageboeck) [reply tree/dataframe/src/RDFHelpers.cxx:182]: Done.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 1 above)
Comment (pcanal) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx]: Still since `fLastPrintTime` can be read unlocked (and thus written to at the same time), it probably should be made an `std::atomic`  (otherwise it can be spuriously be "right".

File: tree/dataframe/src/RDFHelpers.cxx [hunk 11]
```diff
@@ -57,3 +64,161 @@ void ROOT::RDF::RunGraphs(std::vector<RResultHandle> handles)
...
+  const auto eventsPerTimeInterval = currentEventCount - fLastProcessedEvents;
+  fLastProcessedEvents = currentEventCount;
+
+  const auto oldPrintTime = fLastPrintTime;
+  fLastPrintTime = system_clock::now();
+
+  const duration<double> secondsCurrentInterval = fLastPrintTime - oldPrintTime;
+  fEventsPerSecondStatistics[fEventsPerSecondStatisticsIndex++ % fEventsPerSecondStatistics.size()] = eventsPerTimeInterval / secondsCurrentInterval.count();
+
+  return {currentEventCount, duration_cast<seconds>(fLastPrintTime - fBeginTime)};
```
Comment (pcanal) [inline tree/dataframe/src/RDFHelpers.cxx:112]: to reduce contention on "shared" resource (even though the other  access might only be reads).  
```suggestion
  const auto oldPrintTime = fLastPrintTime;
  const auto newPrintTime = system_clock::now();
  fLastPrintTime = system_clock::now();

  const duration<double> secondsCurrentInterval = newPrintTime - oldPrintTime;
  fEventsPerSecondStatistics[fEventsPerSecondStatisticsIndex++ % fEventsPerSecondStatistics.size()] = eventsPerTimeInterval / secondsCurrentInterval.count();

  return {currentEventCount, duration_cast<seconds>(newPrintTime - fBeginTime)};
```

**Comment by guitargeek** [comment]
Superseded by:

* https://github.com/root-project/root/pull/13210
